component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java

```diff
@@ -325,4 +376,97 @@ private void allowUser(String userName, Map claims, X509Certificate cert,
+            String alternativeNamePattern) throws AuthenticationFailedException {
+
+        List<String> matches = new ArrayList<>();
+        try {
+            Collection<List<?>> altNames = cert.getSubjectAlternativeNames();
+            if (altNames != null) {
+                if (alternativeNamesPatternCompiled == null) {
+                    alternativeNamesPatternCompiled = Pattern.compile(alternativeNamePattern);
+                }
+                for (List item : altNames) {
+                    ASN1InputStream decoder = null;
+                    if (item.toArray()[1] instanceof byte[])
+                        decoder = new ASN1InputStream((byte[]) item.toArray()[1]);
+                    else if (item.toArray()[1] instanceof String) {
+                        Matcher m = alternativeNamesPatternCompiled.matcher((String) item.toArray()[1]);
+                        matches = getMatchedString(m, matches);
+                    }
+                    if (decoder == null)
+                        continue;
+                    DEREncodable encoded = decoder.readObject();
+                    encoded = ((DERSequence) encoded).getObjectAt(1);
+                    encoded = ((DERTaggedObject) encoded).getObject();
+                    encoded = ((DERTaggedObject) encoded).getObject();
+                    String identity = ((DERUTF8String) encoded).getString();
+                    Matcher m = alternativeNamesPatternCompiled.matcher(identity);
+                    matches = getMatchedString(m, matches);
+                }
+            } else {
+                authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                        X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_NOTFOUND_ERROR_CODE);
+                throw new AuthenticationFailedException(
+                        X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_NOTFOUND_ERROR);
+            }
+        } catch (CertificateParsingException | IOException e) {
+            throw new AuthenticationFailedException("Failed to Parse the certificate");
+        }
+        if (matches.isEmpty()) {
+            authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                    X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_REGEX_NO_MATCHES_ERROR_CODE);
+            throw new AuthenticationFailedException("Regex Configured but no matches found for the given regex");
+        } else if (matches.size() > 1) {
+            authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                    X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_REGEX_MULTIPLE_MATCHES_ERROR_CODE);
+            throw new AuthenticationFailedException("More than one match for the given regex");
+        } else if (matches.size() == 1) {
+            return matches.get(0);
+        } else {
+            return null;
```

Comment:
-1 for this. Never return null when you have a method signature contract. null creates NPE issues which are harder to handle.

Reply:
done